modulemd/v2/modulemd-yaml-util.c

```diff
@@ -421,3 +441,273 @@ modulemd_yaml_parse_string_set (yaml_parser_t *parser, GError **error)
+                  return FALSE;
+                }
+            }
+          else if (g_str_equal (event.data.scalar.value, "data"))
+            {
+              data = modulemd_yaml_parse_data (parser, &nested_error);
+              if (!data)
+                {
+                  g_propagate_error (error, nested_error);
+                  return FALSE;
+                }
+            }
+
+          break;
+
+        default:
+          MMD_YAML_ERROR_EVENT_EXIT_BOOL (
+            error, event, "Unexpected YAML event in document metadata.");
+        }
+
+      yaml_event_delete (&event);
+    }
+
+  /* The final event must be the document end */
+  YAML_PARSER_PARSE_WITH_EXIT_BOOL (parser, &event, error);
+  if (event.type != YAML_DOCUMENT_END_EVENT)
+    {
+      MMD_YAML_ERROR_EVENT_EXIT_BOOL (
+        error, event, "Document did not end. It just goes on forever...");
+    }
+  yaml_event_delete (&event);
+
+
+  if (doctype == MODULEMD_YAML_DOC_UNKNOWN)
+    {
+      g_set_error_literal (error,
+                           MODULEMD_YAML_ERROR,
+                           MODULEMD_YAML_ERROR_MISSING_REQUIRED,
+                           "No document type specified");
+      return FALSE;
+    }
+
+  if (!mdversion)
+    {
+      g_set_error_literal (error,
+                           MODULEMD_YAML_ERROR,
+                           MODULEMD_YAML_ERROR_MISSING_REQUIRED,
+                           "No metadata version specified");
+      return FALSE;
+    }
+
+  if (!data)
+    {
+      g_set_error_literal (error,
+                           MODULEMD_YAML_ERROR,
+                           MODULEMD_YAML_ERROR_MISSING_REQUIRED,
+                           "No data section provided");
+      return FALSE;
+    }
+
+  *_doctype = doctype;
+  *_mdversion = mdversion;
+  *_data = g_steal_pointer (&data);
+
+  return TRUE;
+}
+
+
+static gchar *
+modulemd_yaml_parse_data (yaml_parser_t *parser, GError **error)
+{
+  MODULEMD_INIT_TRACE ();
+  MMD_INIT_YAML_EVENT (event);
+  MMD_INIT_YAML_EMITTER (emitter);
+  MMD_INIT_YAML_STRING (&emitter, yaml_string);
+  gboolean done = FALSE;
+  gsize depth = 0;
+
+  /* Start the YAML stream */
+  yaml_stream_start_event_initialize (&event, YAML_UTF8_ENCODING);
+  MMD_EMIT_WITH_EXIT_PTR (&emitter, &event, error, "Error starting stream");
+
+  yaml_document_start_event_initialize (&event, NULL, NULL, NULL, 0);
+  MMD_EMIT_WITH_EXIT_PTR (&emitter, &event, error, "Error starting document");
+
+  while (!done)
+    {
+      YAML_PARSER_PARSE_WITH_EXIT (parser, &event, error);
+
+      /* Read in all the YAML from the data section */
+      switch (event.type)
+        {
+        case YAML_SEQUENCE_START_EVENT:
+        case YAML_MAPPING_START_EVENT: depth++; break;
+
+        case YAML_SEQUENCE_END_EVENT: depth--; break;
+
+        case YAML_MAPPING_END_EVENT:
+          depth--;
+          if (depth == 0)
+            done = TRUE;
+          break;
+
+        default: break;
```

Comment:
Perhaps in the `default` case, we should also have a `if (depth == 0) done = TRUE;`, the reason being that otherwise if one has a document with `data: gotcha` or `data: []`, we might be parsing up the rest of the document, and then try to proceed where we should have stopped.
Now, this would most likely be an invalid document further on, but we should probably make sure that we don't fail in unexpected ways.

Maybe do something like:
````
case YAML_SEQUENCE_END_EVENT:
case YAML_MAPPING_END_EVENT:
  depth--
default:
  if (depth == 0)
    done = TRUE;
  break;
````